vertx/src/main/java/com/microsoft/gctoolkit/vertx/jvm/AbstractJavaVirtualMachine.java

```diff
@@ -0,0 +1,150 @@
+ */
+public abstract class AbstractJavaVirtualMachine implements JavaVirtualMachine {
+
+    private static final Logger LOGGER = Logger.getLogger(AbstractJavaVirtualMachine.class.getName());
+
+    private static final double LOG_FRAGMENT_THRESHOLD = 18000;
+
+    private Diarizer jvmConfigurationFromParser;
+    private JvmConfiguration jvmConfigurationForCoreApi;
+    private DateTimeStamp timeOfLastEvent;
+    private final Map<Class<? extends Aggregation>, Aggregation> aggregatedData = new ConcurrentHashMap<>();
+
+    @Override
+    public boolean isG1GC() {
+        return jvmConfigurationFromParser.getDiary().isG1GC();
+    }
+
+    @Override
+    public boolean isZGC() {
+        return jvmConfigurationFromParser.getDiary().isZGC();
+    }
+
+    @Override
+    public boolean isShenandoah() {
+        return jvmConfigurationFromParser.getDiary().isShenandoah();
+    }
+
+    @Override
+    public boolean isParallel() {
+        return jvmConfigurationFromParser.getDiary().isParNew();
+    }
+
+    @Override
+    public boolean isSerial() {
+        return jvmConfigurationFromParser.getDiary().isSerialFull();
+    }
+
+    @Override
+    public boolean isCMS() {
+        return jvmConfigurationFromParser.getDiary().isCMS();
+    }
+
+    @Override
+    public String getCommandLine() {
+        return jvmConfigurationFromParser.getCommandLine();
+    }
+
+    @Override
+    public DateTimeStamp getTimeOfFirstEvent() {
+        return jvmConfigurationFromParser.getTimeOfFirstEvent();
+    }
+
+    @Override
+    public DateTimeStamp getTimeOfLastEvent() {
+        if (getTimeOfFirstEvent().before(timeOfLastEvent))
+            return timeOfLastEvent;
+        else
+            return getTimeOfFirstEvent();
+    }
+
+    @Override
+    public double getRuntimeDuration() {
+        boolean isLogFragment = getTimeOfFirstEvent().getTimeStamp() > LOG_FRAGMENT_THRESHOLD;
+        if (isLogFragment)
+            return getTimeOfLastEvent().minus(getTimeOfFirstEvent());
+        return getTimeOfLastEvent().getTimeStamp();        }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends Aggregation> Optional<T> getAggregation(Class<T> aggregationClass) {
+        return Optional.ofNullable((T) aggregatedData.get(aggregationClass));
+    }
+
+    @Override
+    public JvmConfiguration getJvmConfiguration() {   //todo: why????
+        return jvmConfigurationForCoreApi;
+    }
+
+    // Invoked reflectively from GCToolKit
+    public void analyze(Set<Class<? extends Aggregation>> registeredAggregations, DataSource<?> dataSource) {
+
+        try {
+            final GCLogFile gcLogFile = (GCLogFile) dataSource;
+
+            this.jvmConfigurationFromParser = new UnifiedDiarizer();
+
+            gcLogFile.stream().
+                    filter(Objects::nonNull).
+                    map(String::trim).
+                    filter(s -> s.length() > 0).
+                    map(this.jvmConfigurationFromParser::diarize).
```

Comment:
Another place where it would be good for the diarize method to protect itself from empty/null strings. Would clean up the pipeline a bit.

Reply:
Just another nit... can you put the dot separator at the beginning of the method name instead of at the end of the method call?